### PR TITLE
feat(envs): lazy env init + AsyncVectorEnv as default for n_envs > 1

### PR DESCRIPTION
**Pull request by pkooij**
## Summary

`LiberoEnv` and `MetaWorldEnv` eagerly allocated GPU EGL contexts in `__init__`, making `AsyncVectorEnv` unusable (child processes inherit stale GPU handles → `EGL_BAD_CONTEXT`). All environments were also created upfront, causing OOM on multi-suite evaluations.

This PR:
1. **Defers GPU allocation** to `_ensure_env()`, called on first `reset()`/`step()` inside worker subprocesses
2. **Adds `_LazyAsyncVectorEnv`** — only one task's workers are alive at a time, preventing OOM
3. **Switches default to `AsyncVectorEnv`** for parallel env stepping
4. **Fixes task descriptions** for VLM policies (`env.call("task_description")` instead of broken `add_envs_task`)
5. **Auto-tunes `batch_size`** based on available CPU cores (`batch_size=0`)

## Benchmarks

All runs with `pepijn223/smolvla_libero`, single GPU.

### `libero_spatial` (10 tasks, `batch_size=10`, `n_episodes=10` → 100 rollouts)

| Branch | Wall time | GPU util | GPU mem |
|---|---|---|---|
| `refactor/benchmark-dispatch` | **396s** | 0–8% | ~2 GB |
| `feat/async-vector-env` | **189s** | 0–99% | ~10 GB |

**→ 2.1× speedup**

### Full LIBERO (4 suites, 40 tasks, `n_episodes=10` → 400 rollouts)

| Branch | `batch_size` | Wall time | GPU mem |
|---|---|---|---|
| `refactor/benchmark-dispatch` | 1 (10 OOMs) | **1475s** | ~22 GB |
| `feat/async-vector-env` | 10 | **996s** | ~10 GB |

**→ 1.5× faster, half the GPU memory**

## Related

- Builds on #3272 (benchmark dispatch refactor)

## What changed

- **`libero.py`**: lazy `_ensure_env()` + `_LazyAsyncVectorEnv` wrapper
- **`metaworld.py`**: same lazy init pattern
- **`configs.py`**: default `use_async_envs=True`, auto-downgrade to sync when `n_envs=1`
- **`default.py`**: `batch_size=0` (auto-tune), `use_async_envs=True`
- **`lerobot_eval.py`**: `env.call("task_description")` fix, `env.close()` between tasks
- **`utils.py`**: `_get_sub_env_attr` / `_sub_env_has_attr` for async-compatible attribute access

## Tests

- `test_libero_lazy_init` / `test_metaworld_lazy_init`
- `test_async_vector_env_libero` / `test_async_vector_env_metaworld`
- `test_add_envs_task_async`
- `test_single_env_uses_sync`

### CODE REVIEW

File: src/lerobot/scripts/lerobot_eval.py
Comment (s1lent4gnt) [inline src/lerobot/scripts/lerobot_eval.py]: When using `_LazyAsyncVectorEnv ` which is the default `use_async_envs=True`. Video rendering silently breaks for the most common eval path.

Either make _LazyAsyncVectorEnv inherit from AsyncVectorEnv, or add `else` check.

File: src/lerobot/envs/utils.py [hunk 1]
```diff
@@ -130,56 +130,51 @@ def env_to_policy_features(env_cfg: EnvConfig) -> dict[str, PolicyFeature]:
...
-            raise TypeError(f"Expected task to return a list, got {type(task_result)}")
         if not all(isinstance(item, str) for item in task_result):
             raise TypeError("All items in task result must be strings")
 
         observation["task"] = task_result
-    else:  #  For envs without language instructions, e.g. aloha transfer cube and etc.
+    else:
         num_envs = observation[list(observation.keys())[0]].shape[0]
         observation["task"] = ["" for _ in range(num_envs)]
     return observation
```
Comment (s1lent4gnt) [inline src/lerobot/envs/utils.py]: We can remove this function, it is not used in `lerobot_eval.py`

File: src/lerobot/scripts/lerobot_eval.py
Comment (pkooij) [reply src/lerobot/scripts/lerobot_eval.py]: addressed in https://github.com/huggingface/lerobot/commit/9b131f40

File: src/lerobot/envs/utils.py (hunk 1 above)
Comment (pkooij) [reply src/lerobot/envs/utils.py]: addressed in https://github.com/huggingface/lerobot/commit/97972ae1, only `_get_sub_env_attr` is not used

File: docs/source/env_processor.mdx [hunk 2]
```diff
@@ -126,7 +126,7 @@ class LiberoVelocityProcessorStep(ObservationProcessorStep):
         state = torch.cat([eef_pos, eef_axisangle, eef_vel,
                           gripper_pos, gripper_vel], dim=-1)  # 14D
         return state
-```
+````
```
Comment (s1lent4gnt) [inline docs/source/env_processor.mdx:148]: Please remove this back tick.

File: docs/source/env_processor.mdx
Comment (s1lent4gnt) [inline docs/source/env_processor.mdx:29]: We can remove this.

File: src/lerobot/envs/libero.py [hunk 3]
```diff
@@ -385,6 +403,65 @@ def _make_env(episode_index: int, **kwargs) -> LiberoEnv:
...
+        self._ensure()
+        return self._env.get_attr(name)
+
+    def close(self):
+        if self._env is not None:
+            self._env.close()
+            self._env = None
+
+    def __del__(self):
+        self.close()
```
Comment (s1lent4gnt) [inline src/lerobot/envs/libero.py]: We can remove this and explicitly call `close()` method

File: src/lerobot/scripts/lerobot_eval.py [hunk 4]
```diff
@@ -755,23 +760,39 @@ def _append(key, value):
...
+            if prefetch_thread is not None:
+                prefetch_thread.join()
+                prefetch_thread = None
+
+            # Prefetch next task's AsyncVectorEnv workers while this task runs.
+            if i + 1 < len(tasks):
+                next_env = tasks[i + 1][2]
+                if hasattr(next_env, "_ensure"):
+                    prefetch_thread = threading.Thread(target=next_env._ensure, daemon=True)
+                    prefetch_thread.start()
```
Comment (s1lent4gnt) [inline src/lerobot/scripts/lerobot_eval.py]: We may get GPU memory overlap. Prefetch next task's workers after closing would prevent this, and it's not much loss in performance.

File: src/lerobot/envs/libero.py (hunk 3 above)
Comment (pkooij) [reply src/lerobot/envs/libero.py]: https://github.com/huggingface/lerobot/commit/786ee560

File: src/lerobot/scripts/lerobot_eval.py (hunk 4 above)
Comment (pkooij) [reply src/lerobot/scripts/lerobot_eval.py]: https://github.com/huggingface/lerobot/commit/66d23821

File: src/lerobot/scripts/lerobot_eval.py [hunk 5]
```diff
@@ -166,9 +165,15 @@ def rollout(
...
-        # TODO: works with SyncVectorEnv but not AsyncVectorEnv
-        observation = add_envs_task(env, observation)
+        # Infer "task" from sub-environments (prefer natural language description).
+        # env.call() works with both SyncVectorEnv and AsyncVectorEnv.
+        try:
+            observation["task"] = list(env.call("task_description"))
+        except Exception:
+            try:
+                observation["task"] = list(env.call("task"))
+            except Exception:
```
Comment (s1lent4gnt) [inline src/lerobot/scripts/lerobot_eval.py]: Please catch `text_description` explicitly, `(AttributeError, NotImplementedError)`
We can catch other issues.

Comment (pkooij) [reply src/lerobot/scripts/lerobot_eval.py]: https://github.com/huggingface/lerobot/commit/c3fa286fdd2cba226a71a9663011c78e848e12f6

**Review by s1lent4gnt — APPROVED** [review]
CHAD, LGTM!
